### DISCUSSION TITLE
How long do code completion symbols last when pulled from opened files?

### DISCUSSION THREAD

**Discussion by boocs** [clangd/clangd#1561] [Q&A] [answered]
I noticed that symbols get added to code completion when you open a file that includes the header of the symbols. The opened file isn't in compile_commands.json.

Is there a lifetime to these symbols, for code completion, while the project is open in VSCode? When I close and reopen VSCode the symbols are gone from code completion until I reopen the file that they're included in.

Reason of asking is instead of polluting my compile_commands.json(multi workspace project), I've created a completionHelper.cpp with a bunch of includes of popular header files.

This completionHelper.cpp file isn't part of the project I just open the completionHelper.cpp to load the popular library symbols into clangd's code completion. Now the symbols are available to code completion in any project file.

Initial testing seems to be working.  I wanted to know of any unforeseen problems?

**Comment by HighCommander4** [comment, marked as answer]
Currently I would expect them to stick around through the language server's lifetime.

However, this may be by accident rather than by design. I noticed [this comment](https://searchfox.org/llvm/rev/3947422a77b8f4b191039daac16b9859e0f47f60/clang-tools-extra/clangd/index/FileIndex.h#108) above `FileIndex`:

```c++
/// FIXME: Expose an interface to remove files that are closed.
```

which suggests that in the future, they may go away when `completionHelper.cpp` is closed.

**Reply by boocs** [reply]
Some interesting things:

Let's say we have a symbol called UCameraComponent 

Using the above method. We see the UCameraComponent in code completions.

Adding completionHelper.cpp to compile_commands.json, and not using the above method, doesn't show UCameraComponent. It seems files listed in compile_commands.json are only shallow scanned.

Adding the CameraComponent.cpp to compile_commands.json does show the UCameraComponent symbol for code completion.

**Reply by HighCommander4** [reply]
You may need a clangd restart for clangd to notice the new file in `compile_commands.json` and trigger indexing of it.

**Reply by boocs** [reply]
Yeah, I did a reload window command when testing.

I deleted the cache and restarted VSCode and now it's working. Cool!

**Reply by boocs** [reply]
Some other thing I noticed:

Macros don't load into code completion when the cpp file is in compile_commands.json. Other symbols from the file do though.
Opening the same cpp file does load macros into code completion. 

I'm wondering if it's the reason Unreal does what it does for Intellisense. They preinclude a huge header file for every file.
Note: Unreal Engine requires older clang versions. I'm on clang/clangd 13.0.1

**Reply by boocs** [reply]
Tried doing it the preinclude way but it didn't work. I'll go with opening/loading completionHelper at the start until I figure out something else.
